### PR TITLE
Fix fileMod directory sync for v3

### PR DESCRIPTION
**Pull request by coelho**
On Linux (Ubuntu Server 16.04 at least), fileMod does not handle directory copying properly.

Assume the scenario where "syncdir" is our directory:

```
mkdir test
touch test/a
touch test/b
touch test/c
cp -R test syncdir/test
```

FileMod will notice that "test" has been created, as a directory. It will not notice any of the files _inside_ "test", and as such, none of the files will be synchronized to etcd.

We fix this by, whenever we notice a directory change, iterating through all of the files inside the directory and firing fileMod events for all of them.

This fix is probably in poor taste since, now that I think about it, the fix should be done in main.go's fileMod instead of v3.go. I'm just adding the PR as a proof of concept though towards the issue.

Unrelated note, I really think etcddir should have a "debug" or "verbose" mode to make issues like these easier to find and debug. :heart:

### CODE REVIEW

File: v3.go [hunk 1]
```diff
@@ -127,35 +126,58 @@ func syncProcess_v3(localDir string, serverPrefix string, c3 *clientv3.Client, e
...
+	etcdPath = strings.Replace(etcdPath, string(os.PathSeparator), "/", -1)
+
+	switch {
+	case event.IsRemoved:
+		_, err := c3.Delete(context.Background(), etcdPath)
+		if err != nil {
+			log.Printf("syncProcess_v3 error while delete etcdkey '%v': %v\n", etcdPath, err)
+		}
+	case event.IsDir:
+		files, _ := ioutil.ReadDir(event.Path)
```
Comment (rekby) [inline v3.go:151]: What if the dir have dir's too?

Comment (coelho) [reply v3.go:151]: Should be recursive. @rekby

**Comment by rekby** [comment]
I will try to repeat the problem and think about merge PR or other solve problem, thanks.

File: v3.go (hunk 1 above)
Comment (coelho) [reply v3.go:151]: @rekby With my fix:

```
# mkdir test
# touch test/a
# touch test/b
# touch test/c
# mkdir test/d
# touch test/d/1
# touch test/d/2
# touch test/d/3
# cp -R test dir/test
# /root/etcd/etcdctl get --keys-only=true --prefix=true ""
test/a

test/b

test/c

test/d/1

test/d/2

test/d/3
```

Comment (rekby) [reply v3.go:151]: I don't find recursion in the function. I will try to research it more some later.

File: v3.go [hunk 2]
```diff
@@ -127,35 +126,58 @@ func syncProcess_v3(localDir string, serverPrefix string, c3 *clientv3.Client, e
...
+			path := filepath.Join(event.Path, file.Name())
+			content := []byte(nil)
+			if !file.IsDir() {
+				content, err = ioutil.ReadFile(path)
 				if err != nil {
-					log.Printf("syncProcess_v3 error while put etcdkey '%v': %v\n", etcdPath, err)
+					log.Println(err)
 				}
 			}
+			syncProcess_v3FSEvent(localDir, serverPrefix, c3, fileChangeEvent{
```
Comment (coelho) [inline v3.go:161]: Recursion is right here @rekby  :3

Comment (rekby) [reply v3.go:161]: Yes, I see.
It isn't trivial change, I need some time for research the changes and think about more simple solve the problem.

Comment (coelho) [reply v3.go:161]: Yeah. I just noticed another issue where I did this too:

```
# /root/etcd/etcdctl get --keys-only=true --prefix=true ""
test/a

test/b

test/c

test/d/1

test/d/2

test/d/3

# rm -rf dir/test
# /root/etcd/etcdctl get --keys-only=true --prefix=true ""
test/c

test/d/1
```

And two keys got stuck. I think fileMod really needs to be rechecked in general... It's a bit buggy.

Temporary fixed that by changing:

```
c := make(chan notify.EventInfo, 1)
```

to

```
c := make(chan notify.EventInfo, 256)
```

in main.go, but that is an awful fix. It suits my needs for now though ... haha

**Comment by rekby** [comment]
Yes, I check it too and create auto tests for different scenarios.
Thanks for bugreports.

**Comment by rekby** [comment]
I repeat the problem (and find some others) and create travis-tests.

It can take time to thing about global fix the problem.

**Comment by rekby** [comment]
I have simple tested the fix, it's ok.
Thanks.

**Pull request merged**